File: src/test/java/com/team10/backend/domain/auth/unit/RefreshTokenServiceTest.java
package com.team10.backend.domain.auth.unit;

import com.team10.backend.domain.auth.dto.RefreshResult;
import com.team10.backend.domain.auth.entity.RefreshToken;
import com.team10.backend.domain.auth.repository.RefreshTokenRepository;
import com.team10.backend.domain.auth.service.RefreshTokenService;
import com.team10.backend.domain.user.entity.User;
import com.team10.backend.domain.user.unit.UserTestFixture;
import com.team10.backend.global.exception.BusinessException;
import com.team10.backend.global.exception.ErrorCode;
import com.team10.backend.global.security.TokenProvider;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.time.LocalDateTime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;

@ExtendWith(MockitoExtension.class)
public class RefreshTokenServiceTest {

    @Mock
    private RefreshTokenRepository refreshTokenRepository;

    @Mock
    private TokenProvider tokenProvider;

    @InjectMocks
    private RefreshTokenService refreshTokenService;

    @Test
    @DisplayName("refreshToken 생성 성공")
    void createRefreshToken_success() {
        // given
        User user = UserTestFixture.createBuyer();

        RefreshToken savedToken = mock(RefreshToken.class);

        given(refreshTokenRepository.save(any(RefreshToken.class)))
                .willReturn(savedToken);

        // when
        String result = refreshTokenService.createRefreshToken(user);

        // then
        assertNotNull(result);
        assertFalse(result.isBlank());

        then(refreshTokenRepository).should(times(1)).save(any(RefreshToken.class));
    }


    @Test
    @DisplayName("토큰 재발급 성공")
    void refresh_success() {
        // given
        User user = UserTestFixture.createBuyer();
        ReflectionTestUtils.setField(user, "id", 1L);

        RefreshToken refreshToken = mock(RefreshToken.class);

        given(refreshToken.getUser()).willReturn(user);
        given(refreshToken.getRevoked()).willReturn(false);
        given(refreshToken.getExpiresAt()).willReturn(LocalDateTime.now().plusDays(1));

        given(refreshTokenRepository.findByToken("old-token"))
                .willReturn(refreshToken);

        given(tokenProvider.generateToken(user.getId(), user.getRole()))
                .willReturn("new-access-token");

        given(refreshTokenRepository.save(any(RefreshToken.class)))
                .willReturn(mock(RefreshToken.class));

        // when
        RefreshResult result = refreshTokenService.refresh("old-token");

        // then
        assertEquals("new-access-token", result.accessToken);
        assertNotNull(result.refreshToken);

        then(refreshToken).should().revoke();
        then(refreshTokenRepository).should(times(1)).save(any(RefreshToken.class));
    }

    @Test
    @DisplayName("refreshToken이 null인 경우")
    void refresh_fail_null_token() {
        // when & then
        BusinessException ex = assertThrows(
                BusinessException.class,
                () -> refreshTokenService.refresh(null)
        );

        assertEquals(ErrorCode.MISSING_REFRESH_TOKEN, ex.getErrorCode());
    }

    @Test
    @DisplayName("refreshToken이 없는 값인 경우")
    void refresh_fail_invalid_token() {
        // given
        given(refreshTokenRepository.findByToken("bad-token"))
                .willReturn(null);

        // when & then
        BusinessException ex = assertThrows(
                BusinessException.class,
                () -> refreshTokenService.refresh("bad-token")
        );

        assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.getErrorCode());
    }

    @Test
    @DisplayName("폐기된 토큰인 경우(isRevoked = true)")
    void refresh_fail_revoked_token() {
        // given
        User user = UserTestFixture.createBuyer();

        RefreshToken refreshToken = RefreshToken.create("token", user);
        refreshToken.revoke();

        given(refreshTokenRepository.findByToken("token"))
                .willReturn(refreshToken);

        // when & then
        BusinessException ex = assertThrows(
                BusinessException.class,
                () -> refreshTokenService.refresh("token")
        );

        assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.getErrorCode());
    }

    @Test
    @DisplayName("만료기한이 지난 경우")
    void refresh_fail_expired_token() {
        // given
        RefreshToken refreshToken = mock(RefreshToken.class);

        given(refreshToken.getRevoked()).willReturn(false);
        given(refreshToken.getExpiresAt()).willReturn(LocalDateTime.now().minusDays(1));

        given(refreshTokenRepository.findByToken("token"))
                .willReturn(refreshToken);

        // when & then
        BusinessException ex = assertThrows(
                BusinessException.class,
                () -> refreshTokenService.refresh("token")
        );

        assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.getErrorCode());
    }

}

File: src/test/kotlin/com/team10/backend/domain/auth/unit/RefreshTokenServiceTest.kt
package com.team10.backend.domain.auth.unit

import com.team10.backend.domain.auth.entity.RefreshToken
import com.team10.backend.domain.auth.repository.RefreshTokenRepository
import com.team10.backend.domain.auth.service.RefreshTokenService
import com.team10.backend.fixture.UserFixture
import com.team10.backend.global.exception.BusinessException
import com.team10.backend.global.exception.ErrorCode
import com.team10.backend.global.security.TokenProvider
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.test.util.ReflectionTestUtils
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
internal class RefreshTokenServiceTest {
    @Mock
    lateinit var refreshTokenRepository: RefreshTokenRepository

    @Mock
    lateinit var tokenProvider: TokenProvider

    @InjectMocks
    lateinit var refreshTokenService: RefreshTokenService

    @Test
    @DisplayName("refreshToken 생성 성공")
    fun createRefreshToken_success() {
        // given
        val user = UserFixture.create()

        val savedToken = Mockito.mock<RefreshToken>()

        whenever(refreshTokenRepository.save(any())).thenReturn(savedToken)

        // when
        val result = refreshTokenService.createRefreshToken(user)

        // then
        Assertions.assertNotNull(result)
        Assertions.assertFalse(result.isBlank())

        verify(refreshTokenRepository).save(any())
    }


    @Test
    @DisplayName("토큰 재발급 성공")
    fun refresh_success() {
        // given
        val user = UserFixture.create().apply {
            ReflectionTestUtils.setField(this, "id", 1L)
        }

        val refreshToken = Mockito.mock<RefreshToken>()

        whenever(refreshToken.user).thenReturn(user)
        whenever(refreshToken.revoked).thenReturn(false)
        whenever(refreshToken.expiresAt).thenReturn(
            LocalDateTime.now().plusDays(1)
        )

        whenever(refreshTokenRepository.findByToken("old-token"))
            .thenReturn(refreshToken)

        whenever(tokenProvider.generateToken(user.id, user.role))
            .thenReturn("new-access-token")

        whenever(refreshTokenRepository.save(any()))
            .thenReturn(Mockito.mock<RefreshToken>())

        // when
        val result = refreshTokenService.refresh("old-token")

        // then
        Assertions.assertEquals("new-access-token", result.accessToken)
        Assertions.assertNotNull(result.refreshToken)

        verify(refreshToken).revoke()
        verify(refreshTokenRepository).save(any())
    }

    @Test
    @DisplayName("refreshToken이 null인 경우")
    fun refresh_fail_null_token() {
        // when & then
        val ex =
            assertThrows<BusinessException> { refreshTokenService.refresh(null) }

        Assertions.assertEquals(ErrorCode.MISSING_REFRESH_TOKEN, ex.errorCode)
    }

    @Test
    @DisplayName("refreshToken이 없는 값인 경우")
    fun refresh_fail_invalid_token() {
        // given
        whenever(refreshTokenRepository.findByToken("bad-token"))
            .thenReturn(null)

        // when & then
        val ex =
            assertThrows<BusinessException> { refreshTokenService.refresh("bad-token") }

        Assertions.assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.errorCode)
    }

    @Test
    @DisplayName("폐기된 토큰인 경우(isRevoked = true)")
    fun refresh_fail_revoked_token() {
        // given
        val user = UserFixture.create()

        val refreshToken = RefreshToken.create("token", user)
        refreshToken.revoke()

        whenever(refreshTokenRepository.findByToken("token"))
            .thenReturn(refreshToken)

        // when & then
        val ex =
            assertThrows<BusinessException> { refreshTokenService.refresh("token") }

        Assertions.assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.errorCode)
    }

    @Test
    @DisplayName("만료기한이 지난 경우")
    fun refresh_fail_expired_token() {
        // given
        val refreshToken = Mockito.mock<RefreshToken>()

        whenever(refreshToken.revoked).thenReturn(false)
        whenever(refreshToken.expiresAt).thenReturn(
            LocalDateTime.now().minusDays(1)
        )

        whenever(refreshTokenRepository.findByToken("token"))
            .thenReturn(refreshToken)

        // when & then
        val ex = assertThrows<BusinessException> {
            refreshTokenService.refresh("token")
        }

        Assertions.assertEquals(ErrorCode.INVALID_REFRESH_TOKEN, ex.errorCode)
    }
}
